WIP Deprecate hash key

schneems opened the pull request:
Something like this, need to add a test. Need to extract proc out to a global to avoid object allocation on every call. Move keys back into a hash literal instead of value assignment.

rafaelfranca commented on lib/sprockets/loader.rb:
```diff
@@ -131,16 +131,22 @@ def load_from_unloaded(unloaded)
-            metadata: { dependencies: dependencies }
-          })
+          processor_interface = Hash.new do |hash, key|
+            if :source_path == key
```
It is the other way around. `filepath` is what is deprecated.

schneems replied:
So...I used "filename" in this guide https://github.com/rails/sprockets/blob/master/guides/extending_sprockets.md#supporting-all-versions-of-sprockets-in-processors is it wrong?

In 4.x we have

```ruby
{
            environment: self,
            cache: self.cache,
            uri: unloaded.uri,
            filename: unloaded.filename,
            load_path: load_path,
            source_path: source_path,
            name: name,
            content_type: type,
            metadata: {
              dependencies: dependencies,
              map: []
            }
          }
```

In 3.x we have

```ruby
{
            environment: self,
            cache: self.cache,
            uri: unloaded.uri,
            filename: unloaded.filename,
            load_path: load_path,
            name: name,
            content_type: type,
            metadata: { dependencies: dependencies }
          }
```

It doesn't look like "filepath" was a thing? Also looks like I need to add `source_path` to 3.x. So that. Now I need to figure what exactly `source_path` is in 4.x.

I think a key was changed and removed in a previous version of 3.x by mistake? My compare-view fu is failing me right now.

schneems commented:
As usual i'm more confused now then when I started. The source_path was introduced in 4.x and it represents the path to the source map https://github.com/rails/sprockets/commit/3d54ea834297b7f8456f3c468601cb39e4170fb1


The reason I added that section was because I fixed and submitted a PR to  a repo, I can't find the repo now and naturally didn't link it in the commit message.

schneems commented:
And...the Engines module quit being included into configuration.rb, and didn't have any kind of deprecation notice even on the comments. It's api-breaking turtles all the way down...

schneems commented:
For my own future reference

```
$ be rspec spec
bundler: failed to load command: rspec (/Users/richardschneeman/.gem/ruby/2.3.1/bin/rspec)
NoMethodError: undefined method `register_engine' for #<Sprockets::Environment:0x007f95a0e9e520>
Did you mean?  register_pipeline
```

rafaelfranca commented:
😱 so we should have both filename and source_path?

schneems commented:
Going to close this due to #301
